src/streaming/rules/abr/InsufficientBufferRule.js

```diff
@@ -37,54 +37,79 @@ import SwitchRequest from '../SwitchRequest.js';
 
     let instance,
-        bufferStateDict,
-        lastSwitchTime,
-        waitToSwitchTime;
+        bufferStateDict;
 
     function setup() {
         bufferStateDict = {};
-        lastSwitchTime = 0;
-        waitToSwitchTime = 1000;
         eventBus.on(Events.PLAYBACK_SEEKING, onPlaybackSeeking, instance);
     }
 
+    /*
+     * InsufficientBufferRule does not kick in before the first BUFFER_LOADED event happens. This is reset at every seek.
+     *
+     * If a BUFFER_EMPTY event happens, then InsufficientBufferRule returns switchRequest.quality=0 until BUFFER_LOADED happens.
+     *
+     * Otherwise InsufficientBufferRule gives a maximum bitrate depending on throughput and bufferLevel such that
+     * a whole fragment can be downloaded before the buffer runs out, subject to a conservative safety factor of 0.5.
+     * If the bufferLevel is low, then InsufficientBufferRule avoids rebuffering risk.
+     * If the bufferLevel is high, then InsufficientBufferRule give a high MaxIndex allowing other rules to take over.
+     */
     function getMaxIndex (rulesContext) {
-        let now = new Date().getTime();
         let mediaType = rulesContext.getMediaType();
         let metrics = metricsModel.getReadOnlyMetricsFor(mediaType);
         let lastBufferStateVO = (metrics.BufferState.length > 0) ? metrics.BufferState[metrics.BufferState.length - 1] : null;
         let switchRequest = SwitchRequest(context).create();
 
-        if (now - lastSwitchTime < waitToSwitchTime ||
-            lastBufferStateVO === null) {
+        if (!lastBufferStateVO || !wasFirstBufferLoadedEventTriggered(mediaType, lastBufferStateVO)) {
             return switchRequest;
         }
 
-        setBufferInfo(mediaType, lastBufferStateVO.state);
-        // After the sessions first buffer loaded event , if we ever have a buffer empty event we want to switch all the way down.
-        if (lastBufferStateVO.state === BufferController.BUFFER_EMPTY && bufferStateDict[mediaType].firstBufferLoadedEvent !== undefined) {
+        if (lastBufferStateVO.state === BufferController.BUFFER_EMPTY) {
             log('Switch to index 0; buffer is empty.');
             switchRequest.quality = 0;
             switchRequest.reason = 'InsufficientBufferRule: Buffer is empty';
+        } else {
+            const mediaInfo = rulesContext.getMediaInfo();
+            const abrController = rulesContext.getAbrController();
+            const throughputHistory = abrController.getThroughputHistory();
+            const trackInfo = rulesContext.getTrackInfo();
+            const fragmentDuration = trackInfo.fragmentDuration;
+
+            let bufferLevel = dashMetrics.getCurrentBufferLevel(metrics);
+
+            let throughput = throughputHistory.getAverageThroughput(mediaType);
+            let latency = throughputHistory.getAverageLatency(mediaType);
+
+            let bitrate = throughput * (bufferLevel / fragmentDuration) * INSUFFICIENT_BUFFER_SAFETY_FACTOR;
```

Comment:
> ... the throughput can only ever be halved at most - Was a more radical bitrate change wanted?
I thought 0.5 should be conservative enough, but it's just an arbitrary value which I wouldn't mind changing.

> In the same vein, does this still work for MPDs where fragment duration is not constant?
I don't think fragment duration varies too much. This still works fine with small (<10%) variation in fragment size. Did you have a particular example in mind with larger variation?